content/blog/c-off-ramp.md

```diff
@@ -0,0 +1,218 @@
+
+1. I use the excellent [gnulib](https://www.gnu.org/software/gnulib/) portability library to ensure that my POSIX-compliant code runs on most systems without problems. Unfortunately, gnulib is tied to GNU Autotools. But then, few build systems work on as many platforms as the Autotools in any case.
+2. [Automake](https://www.gnu.org/software/automake/) not only has Vala support, but allows you to ship the generated C sources, so that users can build your code without a Vala compiler. This means that your translated project has exactly the same build dependencies as it had before, which can help with users and packagers who haven’t heard of Vala and don’t want to learn. With Zile, previously a pure-POSIX project, I added GLib as a build-time dependency, but of course that’s so widely used that it’s unlikely to be a problem. Enchant already used GLib.
+
+
+## The Vala experience
+
+I will now describe the process of translating the two code-bases. Although they were a bit different (Zile an application, Enchant a library), and I made the translations a few years apart, I was pleasantly surprised to find the second time that the process was much the same as the first.
+
+### Add Vala to the build system
+
+With Autotools, this is a matter of adding a couple of `configure.ac` lines to detect the Vala compiler and check for GNU Make (which is required to build from Vala sources), and then adding a test Vala source file to a `Makefile.am`.
+
+### Translating the code
+
+I typically translate one file at a time. The code will of course usually call functions from other parts of the program: I add a private VAPI file with suitable bindings. You can usefully pre-populate the VAPI file with the commented-out contents of relevant C header files, then just uncomment the APIs you need in Vala, and edit them suitably. The excellent [Manual bindings](https://wiki.gnome.org/Projects/Vala/ManualBindings) guide from the GNOME Wiki is my constant companion.
+
+To translate a file, I copy `foo.c` to `foo_vala.vala`. (The `_vala` bit ensures that the Vala-generated C won’t overwrite the original file, which I will keep for reference.) I translate the code, possibly in multiple steps, commenting out the bits I have translated in `foo.c`.
+
+Compile, test, fix, rinse, repeat.
+
+I find that I get a nice rhythm with my translation, of doing the obvious basic syntactic stuff, as mentioned above, then finding the equivalent Vala APIs where necessary, then tackling any hard bits, such as binding third-party APIs that Vala does not yet know about.
+
+For Zile I leaned heavily on the POSIX VAPI, and added a new GNU VAPI for some GNU-specific APIs, which I have also used and extended while translating Enchant. Enchant was easier on the whole, because it was already GLib-based, so I just had to search [valadoc.org](https://valadoc.org) for the Vala versions of the GLib APIs.
```

Comment:
I didn't. I would like to investigate this in libpaper where I can get away without GLib data structures (this is a big loss of power to Vala, so only works for either simple code, like libpaper, or where you're prepared to reimplement data structures—e.g. it would be possible to bind lists and hash tables from gnulib).